firefox_puppeteer/api/software_update.py

```diff
@@ -0,0 +1,544 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import ConfigParser
+import os
+import re
+
+import mozinfo
+
+from ..base import BaseLib
+from .appinfo import AppInfo
+from .prefs import Preferences
+
+
+class SoftwareUpdate(BaseLib):
+    """The SoftwareUpdate API adds support for an easy access to the update process."""
+
+    PREF_APP_DISTRIBUTION = 'distribution.id'
+    PREF_APP_DISTRIBUTION_VERSION = 'distribution.version'
+    PREF_APP_UPDATE_URL = 'app.update.url'
+    PREF_DISABLED_ADDONS = 'extensions.disabledAddons'
+
+    def __init__(self, marionette_getter):
+        BaseLib.__init__(self, marionette_getter)
+
+        self.app_info = AppInfo(marionette_getter)
+        self.prefs = Preferences(marionette_getter)
+
+        self.update_channel = UpdateChannel(self)
+        self.mar_channels = MARChannels(self)
+        self.active_update = ActiveUpdate(self)
+
+        self._aus = self.marionette.execute_script("""
+            return Cc['@mozilla.org/updates/update-service;1']
+                .getService(Ci.nsIApplicationUpdateService);
+            """)
+
+    @property
+    def ABI(self):
+        """Get the customized ABI for the update service.
+
+        :returns: ABI version
+        """
+
+        abi = self.app_info.XPCOMABI
+        if mozinfo.isMac:
+            abi += self.marionette.execute_script("""
+                let macutils = Cc['@mozilla.org/xpcom/mac-utils;1']
+                    .getService(Ci.nsIMacUtils);
+                if (macutils.isUniversalBinary)
+                    return '-u-' + macutils.architecturesInBinary;
+
+                return '';
+            """)
+
+        return abi
+
+    @property
+    def allowed(self):
+        """Check if the user has permissions to run the software update
+
+        :returns: Status if the user has the permissions
+        """
+
+        return self._aus['canCheckForUpdates'] and self._aus['canApplyUpdates']
+
+    def assert_update_applied(self, update_data):
```

Comment:
I would suggest we also keep the following order in a class: properties, and then methods. If it's not such a hassle to move some parts around here.